File: lcof-python/ListNode/1_尾到头打印链表.py
# -*- coding:utf-8 -*-
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
#
# 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
#
# 
# @param listNode ListNode类 
# @return int整型一维数组
# -> List[int]
class Solution:
    def printListFromTailToHead(self , listNode: ListNode):
        if listNode is None:
            return []
        a = []
        try:
            while True:
                a.append(listNode.val)
                listNode = listNode.next
        except:
            pass
        return a[::-1]
    
if __name__ == "__main__":
    s = Solution()
    l1 = ListNode(1)
    l2 = ListNode(2)
    l1.next = l2   
    l3 = ListNode(3)    
    l2.next = l3
    l4 = ListNode(4)
    l3.next = l4
    print(s.printListFromTailToHead(l1))

File: lcof-cpp/offer/1_ListNode/1_尾到头打印链表.cpp
#include <bits/stdc++.h>
#include <algorithm>
#include <cstddef>
#include <vector>
struct ListNode
{
    int val;
    struct ListNode *next;
    ListNode(int x) : val(x), next(NULL)
    {
    }
};
class Solution
{
public:
    vector<int> printListFromTailToHead(ListNode *head)
    {
        // 如果头节点为空，直接返回空向量
        if (!head)
            return {};

        // 使用栈来存储节点值
        stack<int> s;
        ListNode *cur = head;

        // 遍历链表并将值压入栈中
        while (cur != nullptr)
        {
            s.push(cur->val);
            cur = cur->next;
        }
        // 创建结果向量，并从栈中弹出元素
        vector<int> v;
        while (!s.empty())
        {
            v.push_back(s.top());
            s.pop();
        }

        return v;
    }
};